sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala

```diff
@@ -348,13 +349,23 @@ object CreateDataSourceTableUtils extends Logging {
         className = provider,
         options = options)
 
-    def newSparkSQLSpecificMetastoreTable(): CatalogTable = {
+    def newSparkSQLSpecificMetastoreTable(relation: HadoopFsRelation): CatalogTable = {
       CatalogTable(
         identifier = tableIdent,
         tableType = tableType,
         schema = Nil,
         storage = CatalogStorageFormat(
-          locationUri = None,
+          // We don't want Hive metastore to implicitly create a table directory,
+          // which may be not the one Data Source table is referring to,
+          // yet which will be left behind when the table is dropped for an external table
+          locationUri = if (isExternal) {
```

Comment:
I think what we can do is to check if the options contains `path`. If so, we put that at here. Otherwise, we leave locationUri as None.